design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+that has not been marked as `Inactive` by its parent.
+
+### Phase 1
+
+_Phase 1 should be completed by release of Crossplane v0.13_
+
+Issues to be Opened:
+
+- [ ] Rename `Package` to `LegacyPackage`
+- [ ] Add `Package` and `PackageRevision` types
+- [ ] Implement `Package` and `PackageRevision` controllers
+- [ ] Deprecate `ClusterPackageInstall`, `PackageInstall`, `StackDefinition`,
+  and `LegacyPackage`
+
+### Phase 2
+
+_Phase 2 should be completed by release of Crossplane v0.14_
+
+Issues to be Opened:
+
+- [ ] Deprecate `app-wordpress`
+- [ ] Update `stack-aws-sample` to use new `Package` format and composition
+- [ ] Update `stack-azure-sample` to use new `Package` format and composition
+- [ ] Update `stack-gcp-sample` to use new `Package` format and composition
+
+`ClusterPackageInstall`, `PackageInstall`, `StackDefinition`, and
+`LegacyPackage` and their corresponding controllers are marked as deprecated.
+All stacks should be migrated to use packages that install composition types.
+
+### Phase 3
+
+_Phase 3 should be completed by release of Crossplane v0.15_
+
+Issues to be Opened:
+
+- [ ] Archive `app-wordpress`
+- [ ] Remove `ClusterPackageInstall`, `PackageInstall`, `StackDefinition`, and
+  `LegacyPackage` and their controllers
+
+### Example Manifests
+
+TODO: this section is just experimenting with YAML formats right now
+
+User creates `Package`:
+
+```yaml
+apiVersion: packages.crossplane.io/v1alpha1
+kind: Package
+metadata:
+  name: provider-gcp
+spec:
+  package: crossplane/provider-gcp:v0.10.0
+```
+
+`Package` creates `Job`:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: provider-gcp
+  namespace: crossplane-system
+  ownerReferences:
+  - apiVersion: packages.crossplane.io/v1alpha1
+    controller: true
+    blockOwnerDeletion: true
+    kind: Package
+    name: provider-gcp
+    uid: d9607e19-f88f-11e6-a518-42010a800195
+spec:
+  template:
+    spec:
+      initContainers:
+      - name: package-copy-to-volume
+        image: crossplane/provider-gcp:v0.10.0
+        cmd: ["cp", "-R", ".registry/", "/ext-pkg/"]
+        volumeMounts:
+        - name: package-contents
+          mountPath: "/ext-pkg"
+      containers:
+      - name: "package-unpack-and-output"
+        args: ["package", "unpack", "--content-dir=/ext-pkg/.registry"]
+        volumeMounts:
+        - name: package-contents
+          mountPath: "/ext-pkg"
+      volumes:
+      - name: package-contents
+        emptyDir: {}
+```
+
+`Package` reads `Job` output and creates `PackageRevision`:
+
+```yaml
+apiVersion: packages.crossplane.io/v1alpha1
+kind: PackageRevision
+metadata:
+  name: cc8a4eb3a83721817ac76c7080622aec8c999538230db98be9755fea66e6982d
+  ownerReferences:
+  - apiVersion: packages.crossplane.io/v1alpha1
+    controller: true
+    blockOwnerDeletion: true
+    kind: Package
+    name: provider-gcp
+    uid: d9607e19-f88f-11e6-a518-42010a800195
+spec:
+  customResourceDefinitions:
```

Comment:
Can you elaborate a little about the purpose of the `customResourceDefinitions`, `infrastructureDefinitions` and `compositions` fields here? Are they essentially records of the things this package packages that other packages may depend on?

It seems like there's a little bit of a discrepancy in the amount of data we provide:

* For CRDs we state "I add support for this group, version and kind"
* For IDs we state "I add support for this group and kind"
* For Compositions we state "I add support for this name"

Reply:
For the CRDs we could update it to just be the name if we wanted as well. Basically this is just so the `PackageRevision` controller can make sure that it is establishing ownership (either by label or explicit owner reference as you are proposing by having the `PackageRevision` manage its own CRDs). Really it is enough to only have whatever uniquely identifies a each in the cluster. We aren't really providing varying levels of information here, it just happens that an ID is typically named as its group kind, and a composition may be named anything.

Reply:
From package manager's perspective, the definition types (`InfrastructureDefinition`, `InfrastructurePublication`, `CustomResourceDefinition`) are actually same; separation from others is because they need to be treated differently but the treatment of the 3 doesn't different between each other. Would it be simpler if we had `definitions` field that contains all these 3 kinds with their GVK. In fact, having `definitions` under one field could help with dependency resolution, too, since IDs/IPs could be depended by other packages just like raw CRDs.

Reply:
@muvaf yeah I think what I was trying to say is that the name of the `InfrastructureDefinition` is the GK. Since they all eventually lead to a CRD though, we could have them all just listed as GVK. However, I continue to struggle with the fact that just because something has the correct GVK, it does not mean that it is the correct CRD. For instance, I could push a package with a CRD with GVK `rdsinstance.aws.crossplane.io/v1beta1` that did not actually represent an AWS RDS instance. If a user installed my package in their cluster, then installed a "stack" (i.e. package with compositions) that depended on a CRD of that GVK (but actually wanted the one from `provider-aws`), the package manager would be say okay I see that CRD exists so this must be okay. This is getting into more of what I mention in this comment: https://github.com/crossplane/crossplane/pull/1616#discussion_r448485301

Reply:
> For instance, I could push a package with a CRD with GVK rdsinstance.aws.crossplane.io/v1beta1 that did not actually represent an AWS RDS instance.

Yeah, this is where the `go.mod` approach makes sense IMO. Think of the `Package` as a Go module; it's one module with many go packages, i.e. CRDs. At the higher level, CRD resolution would prepend the GVKs with module path to decide which one to use. That way, users can choose which module to use for which CRD. But that's all for the CRD resolution. Here at this lower level, the best `Package` resource can do is to have a useful and correct representation of what GVKs it brings into the cluster. So, you suggestion about listing GVKs actually makes sense to me. Something like:
```yaml
types:
- rdsinstance.aws.crossplane.io/v1beta1
- vpc.ec2.aws.crossplane.io/v1beta1
...
```

Reply:
> At the higher level, CRD resolution would prepend the GVKs with module path to decide which one to use.

Are you imagining module path is the package image that provides the CRD?

Reply:
> Are you imagining module path is the package image that provides the CRD?

Yes. Something like `crossplane/provider-aws:v0.10.0/rdsinstance.aws.crossplane.io/v1beta1`